Adds Generic Set type to `utils`

tcsc opened the pull request:
This patch adds a generic set type to the `utils` package, providing
utility methods and a friendlier syntax for the common Go pattern of
modelling a set of `T` as a `map[T]struct{}`.

While having a common implemntation of set operations like `intersection`
is useful, the main motivation for this change is readability. For example,
declaring a `Set[T]` signals the intent of the author much more clearly
than declaring a `map[T]struct{}`.

tcsc committed:
Adds Generic Set type to `utils`

This patch adds a generic set type to the `utils` package, providing
utility methods and a friendlier syntax for the common Go pattern of
modelling a set of `T` as a `map[T]struct{}`.

While having a common implemntation of set operations like `intersection`
is useful, the main motivation for this change is readability. For example,
declaring a `Set[T]` signals the intent of the author much more clearly
than declaring a `map[T]struct{}`.

tcsc commented on lib/utils/set.go:
```diff
@@ -0,0 +1,105 @@
+}
+
+// AddAll expands the set to include all of the supplied elements.
+func (s Set[T]) AddAll(elements []T) {
```
Don't see why not

tcsc replied:
Done

tcsc commented on lib/utils/set.go:
```diff
@@ -0,0 +1,105 @@
+}
+
+// Union updates the set to be the union of the original set and `other`
+func (s Set[T]) Union(other Set[T]) {
```
Yeah, I get that. I just didn't want to create a bunch of temporary sets when chaining operations together. Perhaps it's premature optimization.

bl-nero replied:
I think that since the `Union` doesn't really return a value, the probability of making a mistake is negligible; after all, you'd have to call `Union` and not care about the result. Perhaps consider adding a separate, receiver-less `Union` function that takes two sets and produces a third one, just for convenience. Up to you.

tcsc committed:
Replace Set.AddAll() with variadic Set.Add()

tcsc committed:
Made Remove variadic to match Add()

tcsc committed:
Removed obsolete test, added test for Remove()

tcsc committed:
Merge branch 'master' into tcsc/generic-set

r0mant reviewed (changes requested):
We already have the Set type implemented for predicate: https://github.com/gravitational/teleport/blob/master/lib/expression/set.go.

I'd prefer we avoid having 2 similar yet slightly different implementations of the same thing. If you need a set for IC work, can you factor out/extend/reuse existing implementation?

tcsc commented:
I can probably update the existing set type to use the generic one pretty easily. I'd prefer to do that in a follow-up PR if possible, though.

tcsc committed:
Merge remote-tracking branch 'origin/master' into tcsc/generic-set

tcsc commented:
@r0mant - See #47375 for `expression` refactor

r0mant reviewed (approved):
lgtm assuming https://github.com/gravitational/teleport/pull/47375 merges here first

bl-nero commented on lib/utils/set.go:
```diff
@@ -0,0 +1,102 @@
+
+// Subtract removes all elements in `other` from the set (i.e `s` becomes the
+// Set Difference of `s` and `other`), returning a reference to the mutated set.
+func (s Set[T]) Subtract(other Set[T]) Set[T] {
```
This, on other hand, is going to be confusing: returning the modified set would make it easy to assume that this function doesn't work with place. I'd suggest removing the returned value, and if it's important to have a function that returns something, make a convenience wrapper that returns a clone.

tcsc committed:
Rewrites expression.Set in terms of utils.Set[T] (#47375)

Refactors the Set type in the `expression` package to use `utils.Set[T]` as
the underlying implementations. The `expression.Set` type still exists in
order to implement `expression`-specific interfaces.

tcsc committed:
Merge branch 'master' into tcsc/generic-set

nklaassen commented on lib/expression/set.go:
```diff
@@ -18,65 +18,31 @@
-	return s
-}
-
-func (s Set) add(values ...string) Set {
-	out := s.clone()
-	for _, value := range values {
-		out[value] = struct{}{}
-	}
-	return out
-}
```
The add and remove methods on the generic set mutate the receiver, while this one used to only mutate the copy. Unfortunately it looks like we had zero test coverage for that. I'm going to revert the behaviour on expression.Set to avoid mutating the receiver.

nklaassen replied:
I will revert because I think this could be a subtle breaking change to existing login rules and other expressions that customers have defined

nklaassen replied:
Nevermind, you did the clone inline in the parser spec definition, nothing to see here folks 😅
